Cython/Utility/CythonFunction.c

```diff
@@ -367,37 +367,33 @@ __Pyx_CyFunction_get_annotations(__pyx_CyFunctionObject *op, CYTHON_UNUSED void
     return result;
 }
 
-#if PY_VERSION_HEX >= 0x030402C1
 static PyObject *
 __Pyx_CyFunction_get_is_coroutine(__pyx_CyFunctionObject *op, CYTHON_UNUSED void *context) {
-#if PY_VERSION_HEX >= 0x030503C1
+#if PY_VERSION_HEX >= 0x03050000
     // on v3.5.3rc1 the marker object was introduced
-    if (!op->func_is_coroutine) {
-        if (op->flags & __Pyx_CYFUNCTION_COROUTINE) {
-            PyObject *module, *fromlist, *marker = PYIDENT("_is_coroutine");
-            fromlist = PyList_New(1);
-            if (unlikely(!fromlist)) goto ignore;
-            PyList_SET_ITEM(fromlist, 0, marker);
-            module = PyImport_ImportModuleLevelObject(PYIDENT("asyncio.coroutines"), NULL, NULL, fromlist, 0);
-            Py_DECREF(fromlist);
-            if (unlikely(!module)) goto ignore;
-            op->func_is_coroutine = __Pyx_PyObject_GetAttrStr(module, marker);
-            Py_DECREF(module);
-            if (unlikely(!op->func_is_coroutine)) goto ignore;
-            return op->func_is_coroutine;
-        }
-    ignore:
-        op->func_is_coroutine = Py_None;
+    if (op->func_is_coroutine)
+        return __Pyx_NewRef(op->func_is_coroutine);
+    if (op->flags & __Pyx_CYFUNCTION_COROUTINE) {
+        PyObject *module, *fromlist, *marker = PYIDENT("_is_coroutine");
+        fromlist = PyList_New(1);
+        if (unlikely(!fromlist)) return NULL;
+        PyList_SET_ITEM(fromlist, 0, marker);
+        module = PyImport_ImportModuleLevelObject(PYIDENT("asyncio.coroutines"), NULL, NULL, fromlist, 0);
+        Py_DECREF(fromlist);
+        if (unlikely(!module)) goto ignore;
```

Comment:
This leaves the exception around. Need to call `PyErr_Clear()` somewhere on the way. Same for the other `goto ignore`, in case there was an exception (but we can just always clear it, even if there is none).

Reply:
Ah, I thought your previous comment was just for the GetAttr. I'll revert the NoError usage then, if we're going to clear it anyway.

Reply:
Well, why should we request the creation of an exception first just to clear it afterwards? If we can avoid it right away by calling `…NoError()`, the better.